lago/hooks.py

```diff
@@ -0,0 +1,254 @@
+        os.mkdir(dir)
+        os.mkdir(scripts_dir)
+
+        for cmd, stages in config.viewitems():
+            cmd_dir = path.join(dir, cmd)
+            os.mkdir(cmd_dir)
+            for stage, hooks in stages.viewitems():
+                stage_dir = path.join(cmd_dir, stage)
+                os.mkdir(stage_dir)
+                for hook in hooks:
+                    hook_src_path = path.expandvars(hook)
+                    hook_name = path.basename(hook_src_path)
+                    hook_dst_path = path.join(scripts_dir, hook_name)
+
+                    try:
+                        shutil.copy(hook_src_path, hook_dst_path)
+                        os.symlink(
+                            hook_dst_path, path.join(stage_dir, hook_name)
+                        )
+                    except IOError as e:
+                        raise utils.LagoUserException(e)
+
+
+class Hooks(object):
+
+    PRE_CMD = 'pre'
+    POST_CMD = 'post'
+
+    def __init__(self, path):
+        """
+        Args:
+            path(list of str): path to the hook dir inside the prefix
+
+        Returns:
+            None
+        """
+        self._path = path
+
+    def run_pre_hooks(self, cmd):
+        """
+        Run the pre hooks of cmd
+        Args:
+            cmd(str): Name of the command
+
+        Returns:
+            None
+        """
+        self._run(cmd, Hooks.PRE_CMD)
+
+    def run_post_hooks(self, cmd):
+        """
+        Run the post hooks of cmd
+        Args:
+            cmd(str): Name of the command
+
+        Returns:
+            None
+        """
+        self._run(cmd, Hooks.POST_CMD)
+
+    def _run(self, cmd, stage):
+        """
+        Run the [ pre | post ] hooks of cmd
+        Note that the directory of cmd will be locked by this function in
+        order to avoid circular call, for example:
+
+        a.sh = lago stop
+        b.sh = lago start
+
+        a.sh is post hook of start
+        b.sh is post hook of stop
+
+        start -> a.sh -> stop -> b.sh -> start (in this step the hook
+        directory of start is locked, so start will be called without
+        its hooks)
+
+        Args:
+            cmd(str): Name of the command
+            stage(str): The stage of the hook
+
+        Returns:
+            None
+        """
+        LOGGER.debug('hook called for {}-{}'.format(stage, cmd))
+        cmd_dir = path.join(self._path, cmd)
+        hook_dir = path.join(self._path, cmd, stage)
+
+        if not path.isdir(hook_dir):
+            LOGGER.debug('{} directory not found'.format(hook_dir))
+            return
+
+        _, _, hooks = os.walk(hook_dir).next()
+
+        if not hooks:
+            LOGGER.debug('No hooks were found for command: {}'.format(cmd))
+            return
+
+        # Avoid Recursion
+        try:
+            with utils.DirLockWithTimeout(cmd_dir):
+                self._run_hooks(
+                    sorted([path.join(hook_dir, hook) for hook in hooks])
```

Comment:
I want to preserve the order in which the hooks were mentioned in the init file, but now I realize that I didn't prefix them with an index. will fix.